src/modules/l0-server.c

```diff
@@ -466,30 +492,65 @@ static ksnet_arp_data *ksnLNullSendFromL0(ksnLNullClass *kl, teoLNullCPacket *pa
 int ksnLNullSendToL0(void *ke, char *addr, int port, char *cname,
         size_t cname_length, uint8_t cmd, void *data, size_t data_len) {
 
-    // Create L0 packet
-    size_t out_data_len = sizeof(ksnLNullSPacket) + cname_length + data_len;
-    //char out_data[out_data_len]; // Buffer
-    char *out_data = malloc(out_data_len);
-    memset(out_data, 0, out_data_len);
-    ksnLNullSPacket *spacket = (ksnLNullSPacket*)out_data;
+    int fd = 0;
+    int rv = -1;
+
+    ksnLNullClass *kl = ((ksnetEvMgrClass*)ke)->kl;
+    if (((ksnetEvMgrClass*)ke)->ksn_cfg.l0_allow_f) {
+        fd = ksnLNullClientIsConnected(kl, cname);
+    }
+
+    if (!fd) {
+        //NOTE: client is connected to some other l0 service
+        // Create  Server L0 packet
+        size_t out_data_len = sizeof(ksnLNullSPacket) + cname_length + data_len;
+        //char out_data[out_data_len]; // Buffer
+        char *out_data = malloc(out_data_len);
+        memset(out_data, 0, out_data_len);
+        ksnLNullSPacket *spacket = (ksnLNullSPacket*)out_data;
+
+        // Create teonet L0 packet
+        spacket->cmd = cmd;
+        spacket->from_length = cname_length;
+        memcpy(spacket->from, cname, cname_length);
+        spacket->data_length = data_len;
+        memcpy(spacket->from + spacket->from_length, data, data_len);
+
+        // Send command to client of L0 server
+        #ifdef DEBUG_KSNET
+        ksn_printf(((ksnetEvMgrClass*)ke), MODULE, DEBUG_VV,
+            "send command to L0 server for client \"%s\" ...\n",
+            cname);
+        #endif
+        rv = ksnCoreSendto(((ksnetEvMgrClass*)ke)->kc, addr, port, CMD_L0_TO,
+            out_data, out_data_len);
 
-    // Create teonet L0 packet
-    spacket->cmd = cmd;
-    spacket->client_name_length = cname_length;
-    memcpy(spacket->payload, cname, cname_length);
-    spacket->data_length = data_len;
-    memcpy(spacket->payload + spacket->client_name_length, data, data_len);
+        free(out_data);
+    } else {
+        // Create Client L0 packet
+        size_t out_data_len = sizeof(teoLNullCPacket) + cname_length + data_len;
+        char *out_data = malloc(out_data_len);
+        memset(out_data, 0, out_data_len);
+        size_t packet_length = teoLNullPacketCreate(out_data, out_data_len,
+                cmd, ksnetEvMgrGetHostName((ksnetEvMgrClass*)ke), data, data_len);
 
-    // Send command to client of L0 server
-    #ifdef DEBUG_KSNET
-    ksn_printf(((ksnetEvMgrClass*)ke), MODULE, DEBUG_VV,
-        "send command to L0 server for client \"%s\" ...\n",
-        spacket->payload);
-    #endif
-    int rv = ksnCoreSendto(((ksnetEvMgrClass*)ke)->kc, addr, port, CMD_L0_TO,
-            out_data, out_data_len);
+        // Send command to L0 client
+        if((rv = ksnLNullPacketSend(kl, fd, out_data, packet_length)) >= 0);
```

Comment:
It scares, really. If statement followed semicolon looks suspicious af
could be like 
    rv = ksnLNullPacketSend(kl, fd, out_data, packet_length);
or
    if((rv = ksnLNullPacketSend(kl, fd, out_data, packet_length)) < 0) {
      // log something dirty about "oh-no-we-cant-send-pakkigez!again!"
    }